### PR TITLE
udevadm: make --attribute-walk show subdir attributes

### LINKED ISSUE #12191: udevadm: attribute-walk incomplete - not showing ATTRS below queue/ subdirectory
**systemd version the issue has been seen with**
> systemd 241 (241.67-1-arch)

**Used distribution**
 > Arch Linux 5.0.5-arch1-1-ARCH

**Unexpected behaviour you saw**
 > Some ATTRS not listed

**Steps to reproduce the problem**
`queue/rotational` is not listed:
```
$ udevadm info -a -p /sys/block/sda |grep queue
    ATTRS{queue_ramp_up_period}=="120000"
    ATTRS{queue_type}=="simple"
    ATTRS{queue_depth}=="32"
```
But it's there:
```
$ cat /sys/block/sda/queue/rotational 
1
```
And the rule is working indeed:
```
ACTION=="add|change", KERNEL=="sd[a-z]", ATTR{queue/rotational}=="1", ATTR{queue/scheduler}="bfq"
```
There are many more ATTRS candidates in `ls /sys/block/sda/queue` that can be used as udev rule that don't get caught with the above `udevadm` command.

The udevadm man page says:
```
-a, --attribute-walk
Print all sysfs properties of the specified device that can be used
in udev rules to match the specified device. It prints all devices
along the chain, up to the root of sysfs that can be used in udev
rules.
```
So, shouldn't the above `udevadm` command list all sysfs properties indeed (inclusive `queue/rotational` and `queue/scheduler`)?

### PR DESCRIPTION
**Pull request by yuwata**
Closes #12191.

### CODE REVIEW

File: src/libsystemd/sd-device/sd-device.c [hunk 1]
```diff
@@ -1564,34 +1564,68 @@ _public_ const char *sd_device_get_property_next(sd_device *device, const char *
...
+
                 /* only handle symlinks and regular files */
                 if (!IN_SET(dent->d_type, DT_LNK, DT_REG))
                         continue;
 
-                path = path_join(syspath, dent->d_name);
+                if (subdir)
+                        path = path_join(syspath, subdir, dent->d_name);
+                else
+                        path = path_join(syspath, dent->d_name);
```
Comment (keszybz) [inline src/libsystemd/sd-device/sd-device.c]: `p ?: dent->d_name` could be reused here.

File: src/libsystemd/sd-device/sd-device.c [hunk 2]
```diff
@@ -1564,34 +1564,68 @@ _public_ const char *sd_device_get_property_next(sd_device *device, const char *
...
+                _cleanup_free_ char *p = NULL;
+                struct stat statbuf;
+
+                p = path_join(syspath, subdir, "uevent");
+                if (!p)
+                        return -ENOMEM;
+
+                if (lstat(p, &statbuf) >= 0)
+                        /* this is a child device, skipping */
+                        return 0;
```
Comment (keszybz) [inline src/libsystemd/sd-device/sd-device.c:1587]: Nice!

**Comment by keszybz** [comment]
Downstream udev test fails on bionic-amd64:
```
TEST 30: program with subshell
device '/devices/pci0000:00/0000:00:1f.2/host0/target0:0:0/0:0:0:0/block/sda/sda5' expecting node/link 'bar9'
add:         error
test/tmpfs/dev
├── block
│   └── 8:5 -> ../sda5
├── char
├── null
└── sda5

2 directories, 3 files

remove:      ok
```

Not sure what is going on here.

**Comment by keszybz** [comment]
bionic-i386 timed out, looks unrelated.

File: src/libsystemd/sd-device/sd-device.c (hunk 2 above)
Comment (keszybz) [reply src/libsystemd/sd-device/sd-device.c:1587]: Hmm, actually, why stat this? Isn't `access(p, F_OK)` enough?

File: src/udev/udevadm-info.c [hunk 3]
```diff
@@ -79,14 +107,27 @@ static void print_all_attributes(sd_device *device, const char *key) {
                 if (len > 0)
                         continue;
 
-                printf("    %s{%s}==\"%s\"\n", key, name, value);
+                if (!GREEDY_REALLOC(sysattrs, n_allocated, n_items + 1))
+                        return log_oom();
+
+                sysattrs[n_items].name = name;
+                sysattrs[n_items].value = value;
```
Comment (keszybz) [inline src/udev/udevadm-info.c]: `sysattrs[n_items] = (SysAttr) { ... }` ?

Comment (yuwata) [reply src/udev/udevadm-info.c:114]: Hmm,
```
udevadm-info.c:113:42: error: incompatible types when assigning to type ‘const char *’ from type ‘SysAttr’
  113 |                 sysattrs[n_items].name = (SysAttr) {
      |                                          ^

```

File: src/libsystemd/sd-device/sd-device.c (hunk 2 above)
Comment (keszybz) [reply src/libsystemd/sd-device/sd-device.c:1587]: @yuwata did you see this comment? I don't think we should do stat and throw away the result.

Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:1587]: Oh, sorry. I missed the comment. I will update so. Thank you for the reminder.

**Comment by yuwata** [comment]
@keszybz Updated. Now `access()` is used. PTAL.

**Comment by keszybz** [comment]
LGTM.

File: src/libsystemd/sd-device/sd-device.c [hunk 4]
```diff
@@ -1564,34 +1564,64 @@ _public_ const char *sd_device_get_property_next(sd_device *device, const char *
...
-        dir = opendir(syspath);
+        if (subdir) {
+                _cleanup_free_ char *p = NULL;
+
+                p = path_join(syspath, subdir, "uevent");
+                if (!p)
+                        return -ENOMEM;
+
+                if (access(p, F_OK) >= 0)
+                        /* this is a child device, skipping */
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:1586]: doesn't matter much, but i'd make a more defensive check, and skip things on all errors but ENOENT too. i.e. if there's any doubt better avoid it

Comment (keszybz) [reply src/libsystemd/sd-device/sd-device.c:1586]: I pushed a commit to add that.

**Comment by poettering** [comment]
lgtm

File: src/libsystemd/sd-device/sd-device.c (hunk 4 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:1586]: Thanks.

**Pull request merged**